src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1567,6 +1580,14 @@ public DagState(ImmutableArray<StateForCase> cases, ImmutableDictionary<BoundDag
             /// </summary>
             internal BoundDagTest ComputeSelectedTest()
             {
+                foreach (var item in Cases)
+                {
+                    if (item.RemainingTests is Tests.True or Tests.False)
+                        continue;
+                    var test = item.RemainingTests.ComputeSelectedTest();
+                    if (test.Input.Source is BoundDagPropertyEvaluation { IsLengthOrCount: true })
```

Comment:
:memo: After considering a few alternatives, I think this will do the trick. We just prefer to do the length tests first so that we always know if a indexer access from end could subsume another from beginning.